src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -93,27 +94,18 @@ trait TypeAssigner {
             case _ =>
               mapOver(tp)
           }
+        case tp @ AppliedType(tycon, args) if toAvoid(tycon) =>
+          val base = apply(tycon)
+          val args = tp.baseArgInfos(base.typeSymbol)
+          if (base.typeParams.length == args.length) base.appliedTo(args) else base
```

Comment:
`args` might contain something we want to avoid, this should be `if (base.typeParams.length == args.length) apply(base.appliedTo(args)) else base`.